packages/hash/frontend/src/lib/subgraph.ts

```diff
@@ -194,6 +194,91 @@ export const getEntity = (
   return vertex.inner;
 };
 
+/**
+ * Gets an `Entity` by its `entityId` from within the vertices of the subgraph. Throws an error if
+ * the entity couldn't be found.
+ *
+ * @param params.subgraph
+ * @param params.entityId
+ * @throws if the vertex isn't an `EntityVertex`
+ * @throws if a vertex with `entityId` doesn't exist
+ *
+ * @todo - version is required to identify a specific instance of an entity
+ *   https://app.asana.com/0/1202805690238892/1203214689883091/f
+ */
+export const mustGetEntity = (params: {
+  subgraph: Subgraph;
+  entityId: string;
+}): Entity => {
+  const { entityId, subgraph } = params;
+
+  const entity = getEntity(subgraph, entityId);
+
+  if (!entity) {
+    throw new Error(`Vertex with entityId "${entityId}" does not exist.`);
+  }
+
+  return entity;
+};
+
+/**
+ * Gets the outgoing links of an entity.
+ *
+ * @param params.subgraph
+ * @param params.entityId - the entity id of the source entity.
+ * @param params.linkTypeId (optional) - the id of the link type
+ *
+ * @todo - version is required to identify a specific instance of an entity
+ *   https://app.asana.com/0/1202805690238892/1203214689883091/f
+ */
+export const getOutgoingLinksOfEntity = (params: {
+  entityId: string;
+  subgraph: Subgraph;
+  linkTypeId?: string;
+}): LinkVertex[] => {
+  const { entityId, subgraph, linkTypeId } = params;
+
+  const outgoingLinks = subgraph.edges[entityId]!.filter(
+    ({ edgeKind }) => edgeKind === "HAS_LINK",
+  ).map(({ destination }) => subgraph.vertices[destination] as LinkVertex);
+
+  return linkTypeId
+    ? outgoingLinks.filter(({ inner }) => inner.inner.linkTypeId === linkTypeId)
+    : outgoingLinks;
+};
+
+/**
+ * Gets the incoming links of an entity.
+ *
+ * @param params.subgraph
+ * @param params.entityId - the entity id of the source entity.
+ * @param params.linkTypeId (optional) - the id of the link type
+ *
+ * @todo - version is required to identify a specific instance of an entity
+ *   https://app.asana.com/0/1202805690238892/1203214689883091/f
+ */
+export const getIncomingLinksOfEntity = (params: {
+  entityId: string;
+  subgraph: Subgraph;
+  linkTypeId?: string;
+}): LinkVertex[] => {
+  const { entityId, subgraph, linkTypeId } = params;
+
+  /** @todo: return the incoming links of an entity in a more efficient representation */
+  const incomingLinks = Object.entries(subgraph.edges)
+    .filter(([_, linkOutgoingEdges]) =>
+      linkOutgoingEdges.some(
+        ({ edgeKind, destination }) =>
+          edgeKind === "HAS_DESTINATION" && destination === entityId,
+      ),
+    )
+    .map(([source]) => subgraph.vertices[source] as LinkVertex);
+
+  return linkTypeId
+    ? incomingLinks.filter(({ inner }) => inner.inner.linkTypeId === linkTypeId)
+    : incomingLinks;
```

Comment:
> it'd still be unclear if you should constrain by linkTypeId first, or still search for all incoming links

I think this is a problem we'll need to confront at some point, but it'd be ideally encapsulated in a method like this. I'm happy to add a `@todo`, but my instinct is to encapsulate temporary implementations like this rather than repeating them throughout the codebase wherever these methods are called?

Reply:
> encapsulate temporary implementations like this rather than repeating them throughout the codebase wherever these methods are called?

I think that's fair, and should help us identify which functions would be helpful in some common-library.

> I think this is a problem we'll need to confront at some point

I also agree, I just think that the more simple we keep the methods, the more control callers of the functions will have around what they're optimising for (like they may _know_ that filtering by a linkTypeId first will make it a lot faster). If we shove that inside an implementation they may not be able to control it and therefore have to recreate a bunch more logic.